### PR TITLE
JAMES-3817 Health Check for RabbitMQ dead letter queues

### PR DESCRIPTION
**Pull request by quantranhong1999**
Healthcheck for:
- MailQueue: dead letter queue
- EventBus: MailboxEvent + JmapEvent dead letter queues

### CODE REVIEW

File: event-bus/distributed/src/main/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheck.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public ComponentName componentName() {
+        return COMPONENT_NAME;
+    }
+
+    @Override
+    public Mono<Result> check() {
+        return Mono.fromCallable(() -> api.queueDetails(configuration.getVhost().orElse(DEFAULT_VHOST), mailboxEventNamingStrategy.deadLetterQueue().getName()).getQueueLength())
```
Comment (chibenwa) [inline event-bus/distributed/src/main/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheck.java:54]: blocking calls shall be subscribed on the boundedElastic scheduler or bad things will happen.

https://github.com/reactor/BlockHound is a good tool to find such class of problems...

File: event-bus/distributed/src/test/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheckTest.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.google.common.collect.ImmutableMap;
+import com.rabbitmq.client.AMQP;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ConnectionFactory;
+
+class RabbitMQEventBusDeadLetterHealthCheckTest {
```
Comment (chibenwa) [inline event-bus/distributed/src/test/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheckTest.java:50]: What happens when I have no dead-letter queue (edge case)

File: event-bus/distributed/src/test/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheckTest.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            .build();
+
+        channel.basicPublish(EXCHANGE_NAME, routingKey, basicProperties, "Hello, world!".getBytes(StandardCharsets.UTF_8));
+    }
+
+    private void closeQuietly(AutoCloseable... closeables) {
+        Arrays.stream(closeables).forEach(this::closeQuietly);
+    }
+
+    private void closeQuietly(AutoCloseable closeable) {
```
Comment (chibenwa) [inline event-bus/distributed/src/test/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheckTest.java:143]: There's likely a method in IOUtils to do just this...

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueDeadLetterHealthCheck.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.james.backends.rabbitmq.RabbitMQConfiguration;
+import org.apache.james.backends.rabbitmq.RabbitMQManagementAPI;
+import org.apache.james.core.healthcheck.ComponentName;
+import org.apache.james.core.healthcheck.HealthCheck;
+import org.apache.james.core.healthcheck.Result;
+
+import reactor.core.publisher.Mono;
+
+public class RabbitMQMailQueueDeadLetterHealthCheck implements HealthCheck {
+    private static final ComponentName COMPONENT_NAME = new ComponentName("RabbitMQMailQueueHealthCheck");
```
Comment (chibenwa) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueDeadLetterHealthCheck.java:33]: RabbitMQMailQueueDeadLetterQueueHealthCheck

File: event-bus/distributed/src/main/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheck.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.james.backends.rabbitmq.RabbitMQConfiguration;
+import org.apache.james.backends.rabbitmq.RabbitMQManagementAPI;
+import org.apache.james.core.healthcheck.ComponentName;
+import org.apache.james.core.healthcheck.HealthCheck;
+import org.apache.james.core.healthcheck.Result;
+
+import reactor.core.publisher.Mono;
+
+public class RabbitMQEventBusDeadLetterHealthCheck implements HealthCheck {
+    private static final ComponentName COMPONENT_NAME = new ComponentName("RabbitMQEventBusDeadLetterHealthCheck");
```
Comment (chibenwa) [inline event-bus/distributed/src/main/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheck.java:33]: RabbitMQEventBusDeadLetterQueueHealthCheck

File: event-bus/distributed/src/test/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheckTest.java (hunk 3 above)
Comment (quantranhong1999) [reply event-bus/distributed/src/test/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheckTest.java]: I tried and it does not work in this case. 
`IOUtils.closeQuietly(Closable ...)`: can not down casting rabbitmq reactor `Channel` type' `AutoClosable` to `Closable`.

File: event-bus/distributed/src/test/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheckTest.java (hunk 2 above)
Comment (quantranhong1999) [reply event-bus/distributed/src/test/java/org/apache/james/events/RabbitMQEventBusDeadLetterHealthCheckTest.java]: I suppose it should return unhealthy (in case the operator by mistake deletes the dead-letter queue which should exist to store dead-letter messages).

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/DistributedHealthCheckIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.james.modules.blobstore.BlobStoreConfiguration;
+import org.apache.james.queue.rabbitmq.RabbitMQMailQueueDeadLetterQueueHealthCheck;
+import org.apache.james.utils.GuiceProbe;
+import org.apache.james.webadmin.integration.HealthCheckProbe;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.google.inject.multibindings.Multibinder;
+
+class DistributedHealthCheckIntegrationTest {
```
Comment (chibenwa) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/DistributedHealthCheckIntegrationTest.java:50]: Can't we do this within `WebAdminServerIntegrationImmutableTest` ?

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/DistributedHealthCheckIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Test
+    void shouldContainRabbitMQDeadLetterQueueHealthChecks(GuiceJamesServer server) {
+        List<ComponentName> healthCheckComponentNames = server.getProbe(HealthCheckProbe.class).getHealthCheckSet()
+            .stream()
+            .map(HealthCheck::componentName)
+            .collect(Collectors.toList());
+
+        assertThat(healthCheckComponentNames).contains(RabbitMQEventBusDeadLetterQueueHealthCheck.COMPONENT_NAME,
+            RabbitMQMailQueueDeadLetterQueueHealthCheck.COMPONENT_NAME);
```
Comment (chibenwa) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/DistributedHealthCheckIntegrationTest.java:80]: Can't we list all healtch checks?

Comment (chibenwa) [reply server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/DistributedHealthCheckIntegrationTest.java:80]: Maybe specify hard coded names and specify raw strings?

It will help us catch breaking changes!

File: server/protocols/webadmin-integration-test/memory-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/memory/MemoryHealthCheckIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Test
+    void shouldNotContainRabbitMQDeadLetterQueueHealthChecks(GuiceJamesServer server) {
+        List<ComponentName> healthCheckComponentNames = server.getProbe(HealthCheckProbe.class).getHealthCheckSet()
+            .stream()
+            .map(HealthCheck::componentName)
+            .collect(Collectors.toList());
+
+        assertThat(healthCheckComponentNames).doesNotContain(new ComponentName("RabbitMQMailQueueDeadLetterQueueHealthCheck"),
+            new ComponentName("RabbitMQEventBusDeadLetterQueueHealthCheck"));
```
Comment (chibenwa) [inline server/protocols/webadmin-integration-test/memory-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/memory/MemoryHealthCheckIntegrationTest.java:64]: Or do a contains only on the checks for memory?

File: server/protocols/webadmin-integration-test/memory-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/memory/MemoryHealthCheckIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            .usersRepository(DEFAULT)
+            .build())
+        .server(configuration -> MemoryJamesServerMain.createServer(configuration)
+            .overrideWith(binder -> Multibinder.newSetBinder(binder, GuiceProbe.class).addBinding().to(HealthCheckProbe.class))
+            .overrideWith(new TestJMAPServerModule()))
+        .build();
+
+    @Test
+    void shouldNotContainRabbitMQDeadLetterQueueHealthChecks(GuiceJamesServer server) {
+        List<ComponentName> healthCheckComponentNames = server.getProbe(HealthCheckProbe.class).getHealthCheckSet()
```
Comment (chibenwa) [inline server/protocols/webadmin-integration-test/memory-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/memory/MemoryHealthCheckIntegrationTest.java:58]: Can't we call webadmin instead?

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/DistributedHealthCheckIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        .extension(new AwsS3BlobStoreExtension())
+        .extension(new RabbitMQExtension())
+        .server(configuration -> CassandraRabbitMQJamesServerMain.createServer(configuration)
+            .overrideWith(binder -> Multibinder.newSetBinder(binder, GuiceProbe.class).addBinding().to(HealthCheckProbe.class))
+            .overrideWith(new TestJMAPServerModule()))
+        .build();
+
+    @Test
+    void shouldContainRabbitMQDeadLetterQueueHealthChecks(GuiceJamesServer server) {
+        List<ComponentName> healthCheckComponentNames = server.getProbe(HealthCheckProbe.class).getHealthCheckSet()
```
Comment (chibenwa) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/DistributedHealthCheckIntegrationTest.java:74]: Can't we call webadmin instead?

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]
